drivers/adc/adc_gd32.c

```diff
@@ -0,0 +1,403 @@
+	}
+
+	/* Signle conversion mode with regular group. */
+	ADC_RSQ2(cfg->reg) &= ~ADC_RSQX_RSQN;
+	ADC_RSQ2(cfg->reg) = index;
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_gd32_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_gd32_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_gd32_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, true, async);
+	error = adc_gd32_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static struct adc_driver_api adc_gd32_driver_api = {
+	.channel_setup = adc_gd32_channel_setup,
+	.read = adc_gd32_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_gd32_read_async,
+#endif /* CONFIG_ADC_ASYNC */
+};
+
+static int adc_gd32_init(const struct device *dev)
+{
+	struct adc_gd32_data *data = dev->data;
+	const struct adc_gd32_config *cfg = dev->config;
+	int ret;
+
+	data->dev = dev;
+
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+	/* Select adc clock source and its prescaler. */
+	rcu_adc_clock_config(cfg->rcu_clock_source);
+#endif
+
+	rcu_periph_clock_enable(cfg->rcu_periph_clock);
+
+#if defined(CONFIG_SOC_SERIES_GD32F403) || \
+	defined(CONFIG_SOC_SERIES_GD32VF103) || \
+	defined(CONFIG_SOC_SERIES_GD32F3X0)
+	/* Set SWRCST as the regular channel external trigger. */
+	ADC_CTL1(cfg->reg) &= ~ADC_CTL1_ETSRC;
+	ADC_CTL1(cfg->reg) |= CTL1_ETSRC(7);
+
+	/* Enable external trigger for regular channel. */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ETERC;
+#endif
+
+	/* Enable ADC */
+	ADC_CTL1(cfg->reg) |= ADC_CTL1_ADCON;
+
+	adc_gd32_calibration(cfg);
+
+	cfg->irq_config_func();
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+#define HANDLE_IRQS(idx)							\
+	static const struct device *dev_##idx = DEVICE_DT_INST_GET(idx);	\
+	const struct adc_gd32_config *cfg_##idx = dev_##idx->config;		\
+										\
+	if (ADC_CTL0(cfg_##idx->reg) & ADC_CTL0_EOCIE) {			\
+		adc_gd32_isr(dev_##idx);					\
+	}
+
+static void adc_gd32_shared_irq_handler(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_gd32_shared_irq_configure(void)
+{
+	/* Use a flag to protect the shared irq handler initialization. */
+	static bool init_irq = true;
+
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    adc_gd32_shared_irq_handler,
+			    NULL, 0);
+		irq_enable(DT_INST_IRQN(0));
```

Comment:
this assume information from instance 0 is always the correct one, is that always the case?

Reply:
Correct for most of gd32 soc.

As I know there have a exception for gd32f403.  gd32f403 have 3 ADC. ADC0 and ADC1 sharing the interrupt 18, but ADC3 use an individual interrupt 47.

I have tried several way to fix gd32f403 ADC2. The main goal is fix it at preprocess times:
  - use DTS macro, like DT_FOREACH;
  - use DTS property, like `sharing-interrupts`.

Finally, I'm failed. So I just ignore the ADC2, gd32f403 DTS only have ADC0 and ADC2 now.

Reply:
The irq sharing will only work when all of the ADC share the same interrupt line.
You might be able to workaround this with hardcoded `#ifdef` if there is not a lot of combinations, otherwise you will have the same issue as #39565 which no one has a fix yet.

Reply:
After looks all gd32 series user manual, I get below ADC interrupt table.

|    GD32 SoC    |    ADC0 IRQN    |    ADC1 IRQN    |    ADC2 IRQN    |
|     -----------       |       -----------        |    -----------          |       -----------        |
|  GD32F3x0      |
|  GD32E23x      |
|  GD32E23x      |